### PR TITLE
Feature: Display block conent on course page

### PR DESCRIPTION
**Pull request by dasistwas**
Hi David,
this is the result of the devcamp of MoodleMoot DACH 2019. Thank you for your great support. I hope you are happy with the new feature and can merge it. 
I am looking forward to your feedback.
Kind regards,
David

### CODE REVIEW

File: classes/poster.php [hunk 1]
```diff
@@ -0,0 +1,118 @@
+<?php
+// This file is part of mod_datalynx for Moodle - http://moodle.org/
```
Comment (mudrd8mz) [inline classes/poster.php]: There are a few mentions of datalynx in the patch.

File: classes/poster.php [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+// It is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ *
```
Comment (mudrd8mz) [inline classes/poster.php]: I usually put something like "Provides the {@link mod_poster\poster} class." here.

File: classes/poster.php [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+ * @package mod_poster
+ * @copyright based on the work by 2015 David Mudrak <david@moodle.com>{
```
Comment (mudrd8mz) [inline classes/poster.php]: Extra trailing curly bracket.

File: classes/poster.php [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+namespace mod_poster;
+
+use mod_poster\event\course_module_viewed;
+use completion_info;
+
+defined('MOODLE_INTERNAL') or die();
+
+/**
+ * Class poster
```
Comment (mudrd8mz) [inline classes/poster.php]: This does not add any information, it is a useless comment. The purpose of the class should be described here, e.g. what it represents.

File: classes/poster.php [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+ *
+ * @package mod_datalynx
```
Comment (mudrd8mz) [inline classes/poster.php]: datalynx

File: classes/poster.php [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+ */
+class poster {
+
+    /**
+     * @var mixed|object|false fetched from get_record.
```
Comment (mudrd8mz) [inline classes/poster.php]: This should be just `@var object|false` (which I prefer) or `@var mixed` but not both imho

File: classes/poster.php [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+        global $DB;
+        $this->settings = $DB->get_record('poster', array('id' => $cm->instance), '*', MUST_EXIST);
+        $this->course = $DB->get_record('course', array('id' => $cm->course), '*', MUST_EXIST);
+        $this->cm = $cm;
+    }
+
+    /**
+     * Get settings saved in DB for the poster module.
+     *
+     * @return false|mixed|object
```
Comment (mudrd8mz) [inline classes/poster.php]: Again, I would not mix "mixed" here, it is not a valid type

File: classes/poster.php [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+     */
+    public function get_settings(){
+        return $this->settings;
+    }
+
+    /**
+     * Get course settings.
+     *
+     * @return false|mixed|object
```
Comment (mudrd8mz) [inline classes/poster.php]: mixed

File: db/upgrade.php [hunk 9]
```diff
@@ -33,8 +33,20 @@
  */
 function xmldb_poster_upgrade($oldversion) {
     global $CFG, $DB;
-
     $dbman = $DB->get_manager();
 
+    if ($oldversion < 2019091100) {
+        // Add new fields to poster table.
```
Comment (mudrd8mz) [inline db/upgrade.php]: The plural is confusing, we add a single field here. Can we have it like "Add a new field 'display' to the poster table."

File: lang/en/poster.php [hunk 10]
```diff
@@ -24,7 +24,10 @@
  */
 
 defined('MOODLE_INTERNAL') || die();
-
```
Comment (mudrd8mz) [inline lang/en/poster.php:27]: Can we leave the empty line here please

File: lib.php [hunk 11]
```diff
@@ -148,3 +152,91 @@ function poster_page_type_list($pagetype, $parentcontext, $currentcontext) {
...
+        'id, name, display, showdescriptionview, intro, introformat'))) {
+        return null;
+    }
+    $cminfo = new cached_cm_info();
+    $cminfo->name = $poster->name;
+    if ($poster->display == POSTER_DISPLAY_INLINE) {
+        // prepare poster object to store in customdata
+        $fdata = new stdClass();
+        $fdata->showdescriptionview = $poster->showdescriptionview;
+        if ($cm->showdescription && strlen(trim($poster->intro))) {
```
Comment (mudrd8mz) [inline lib.php]: Better to use `html_is_blank()` here, consistently with how the `view_page_description()` method does it

File: lib.php [hunk 12]
```diff
@@ -148,3 +152,91 @@ function poster_page_type_list($pagetype, $parentcontext, $currentcontext) {
...
+ * if poster.display == POSTER_DISPLAY_INLINE
+ *
+ * @param cm_info $cminfo
+ */
+function poster_cm_info_view(cm_info $cminfo) {
+    global $PAGE;
+    $poster = new poster($cminfo);
+
+    // if ($cm->uservisible && $cm->customdata &&
+    // has_capability('mod/poster:view', $cm->context)) {
```
Comment (mudrd8mz) [inline lib.php]: Forgotten commented lines?

File: lib.php [hunk 13]
```diff
@@ -148,3 +152,91 @@ function poster_page_type_list($pagetype, $parentcontext, $currentcontext) {
...
+    if ($cminfo->uservisible && $cminfo->customdata && has_capability('mod/poster:view', $cminfo->context)) {
+        // Restore poster object from customdata.
+        // Note the field 'customdata' is not empty IF AND ONLY IF we display contens inline.
+        // Otherwise the content is default.
+        // $poster = $cm->customdata;
+        // $poster->id = (int)$cm->instance;
```
Comment (mudrd8mz) [inline lib.php]: Forgotten commented lines?

File: mod_form.php [hunk 14]
```diff
@@ -51,6 +51,16 @@ public function definition() {
...
+        // Add the option to display the content on the course page.
+        $mform->addElement('select', 'display', get_string('display', 'mod_poster'),
+            array(POSTER_DISPLAY_PAGE => get_string('displaypage', 'mod_poster'),
+                POSTER_DISPLAY_INLINE => get_string('displayinline', 'mod_poster')));
+        $mform->addHelpButton('display', 'display', 'mod_poster');
+        if (!$this->courseformat->has_view_page()) {
+            $mform->setConstant('display', POSTER_DISPLAY_PAGE);
+            $mform->hardFreeze('display');
+        }
+
```
Comment (mudrd8mz) [inline mod_form.php]: I would move the field elsewhere in the form. The original idea was to have the "Display name on view page" right below the "Name" field. I would move this setting to the very bottom of the General section probably.

File: renderer.php [hunk 15]
```diff
@@ -114,34 +114,34 @@ protected function view_page_content($poster) {
         $cssclassmain .= $haspre ? '' : ' empty-region-mod_poster-pre';
         $cssclassmain .= $haspost ? '' : ' empty-region-mod_poster-post';
 
-        $out .= html_writer::start_div($cssclassmain, array('id' => 'mod_poster-content'));
+        $out .= \html_writer::start_div($cssclassmain, array('id' => 'mod_poster-content'));
```
Comment (mudrd8mz) [inline renderer.php]: Why is this change needed? I prefer not to have these backslashes in non-namespaced files personally.

File: renderer.php [hunk 16]
```diff
@@ -114,34 +114,34 @@ protected function view_page_content($poster) {
...
 
         // The bootstrap3 based themes should use the class .row here.
         // But that would have different meaning in the bootstrap2 based themes.
         // However, .row-fluid seems to work in bootstrap3 too so leaving it here for now.
-        $out .= html_writer::start_div('row-fluid');
+        $out .= \html_writer::start_div('row-fluid');
```
Comment (mudrd8mz) [inline renderer.php]: dtto

Comment (mudrd8mz) [reply renderer.php]: and all the other changes in this file

**Review by mudrd8mz — CHANGES_REQUESTED** [review]
Many thanks David for this nice feature. I found it working well. It will be a great addition to the module.

I like the overall design and implementation of the feature. I only left a few notes in the patch which I think could be looked at again.

Additionally, would you like to add some Behat tests covering this feature?

**Comment by mudrd8mz** [comment]
Hi @dasistwas - just checking to know if you are considering to work on the suggested fixes yourself, of whether I should do them. I'd still like to have your feature included in the Poster. Thanks for letting me know.

**Comment by dasistwas** [comment]
Hi @mudrd8mz,
thank you for your reminder. I will try to do that tonight. Current situation: Kids at home and still a lot of tasks to do, so I try to catch up when kids are sleeping.
Kind regards,
David

**Comment by mudrd8mz** [comment]
No stress at all, it was not meant as a pushing. I just wanted to clarify so that this does not fall off.

**Comment by dasistwas** [comment]
Hi David,
I resolved all the issues from your code analysis. Thank you for the detailed code review. (I did not see all the comments, because some were hidden, so I just amended the things I forgot to solve). Concerning behat tests: I will put that on my agenda, but I do not think, that this week I will get them done.

**Comment by mudrd8mz** [comment]
> I did not see all the comments, because some were hidden

What do you mean by that?

**Comment by dasistwas** [comment]
Github did collapse some comments like that, which I did not see immediatly:
Screenshot_2020-03-24 Feature Display block conent on course page by dasistwas · Pull Request #9 · mudrd8mz moodle-mod_poster

**Comment by mudrd8mz** [comment]
Ah, ok. Don't worry about Behat.

**Comment by mudrd8mz** [comment]
Hi @dasistwas 

Thanks for the update. I am playing with the feature and I am experiencing some weird behaviour that I think might easily confuse users.

When the Poster is set to display blocks on the course page, and the editing mode is turned on:
* I noticed that clicking the "move block" widgets lead to PHP notices due to missing regions in the list returned by the theme. I discussed this with Tim Hunt in the chat and it seems that this is a Moodle core bug which could be fixed by https://gist.github.com/mudrd8mz/4771f0a1a18918050575214ae1e06958
* However, when using the new feature with the Classic theme, somehow the Navigation and the Administration blocks were added to the Poster region and I can't get rid of them now. I have no clue why this is happening.
* Additionally, I noticed that the behaviour of some blocks can become a little bit tricky. For example the Comments block display the comments associated with the current page context. So it displays different comments when shown on the course page and when shown on the a separate view.php page. This may be unexpected behaviour for teachers who change the Poster display setting there and back again.

May I ask you to try to reproduce the situation with the Classic theme? I am really worried that this could easily break people's Posters if they use a non-Boost theme...

**Comment by mudrd8mz** [comment]
I am closing this feature request for now. I am afraid the situation with the behaviour described above (different page context depending on where the blocks are displayed) would be too confusing and hard predictable for end users. Even though I liked the idea very much at the beginning, I now tend to think that the module should better support only blocks on its own view page only.

**Comment by dasistwas** [comment]
Hi @mudrd8mz,
I do understand your decision. I did not have time to work on that. Some problems you encountered might be solved, others stay tricky:
1) Regions problem: The problem only occurs in editing mode so, the basic idea is to not display the blocks in editing mode, but provide a link to edit the blocks and then you are editing them in the usual way.
2) Classic theme: I will have to check that. But it seems it's a child theme of boost.
3) Context: That might be the most tricky thing. I noticed that as well. So displaying blocks on the course page will have context course and displaying them on the modules page, will have context module. Not sure how to deal with that.

For me the whole block thing would be something to change with Moodle 4.0 UX improvements. Blocks in Moodle have major usability issues, that increased with the implementation of the boost theme and the moodle app.

Hope I can find some time to work on this again. 
Kind regards,
David

**Comment by mudrd8mz** [comment]
Thanks David @dasistwas 

Re 3., maybe there could be a switch so that the teacher could decide whether they intent to display the blocks on the course page or a special view page. And we would just not support both for consistency.

That might work imho.

**Pull request closed without merging**